server/models/pattern/core/pattern.go

```diff
@@ -325,46 +325,84 @@ func processCytoElementsWithPattern(eles []cytoscapejs.Element, callback func(sv
+// and returns a slice of resource manifests as generic map[string]interface{} structures.
+func getAllK8sResources(data string, ignoreErrors bool) ([]map[string]interface{}, error) {
+	var K8sResources []map[string]interface{}
+	decoder := yaml.NewDecoder(bytes.NewBufferString(data))
 
-	buffer := bytes.NewBufferString(data)
-	decoder := yaml.NewDecoder(buffer)
 	for {
 		manifest := map[string]interface{}{}
-
 		err := decoder.Decode(manifest)
 		if err != nil {
 			if err == io.EOF {
-				if len(pattern.Components) == 0 {
-					return pattern, ErrParseK8sManifest(fmt.Errorf("kubernetes manifest is empty"))
-				}
-				return pattern, nil
-			} else {
-				return pattern, ErrParseK8sManifest(fmt.Errorf("kubernetes manifest is empty"))
+				break
 			}
+			return K8sResources, ErrParseK8sManifest(fmt.Errorf("Kubernetes manifest is invalid: %v", err))
 		}
+
+		// Skip empty manifests
 		if len(manifest) == 0 {
 			continue
 		}
-		// Recursive casting
+
 		manifest = utils.RecursiveCastMapStringInterfaceToMapStringInterface(manifest)
 		if manifest == nil {
 			if ignoreErrors {
 				continue
 			}
-			return pattern, ErrParseK8sManifest(fmt.Errorf("failed to parse manifest into an internal representation"))
+			return K8sResources, ErrParseK8sManifest(fmt.Errorf("failed to parse Kubernetes manifest into an internal representation"))
 		}
 
+		// If it's a List kind, expand it, otherwise add directly
+		kind, _ := mutils.Cast[string](manifest["kind"])
+		if kind == "List" {
+			if items, ok := manifest["items"].([]interface{}); ok {
+				for _, item := range items {
+					if itemMap, ok := item.(map[string]interface{}); ok {
+						K8sResources = append(K8sResources, itemMap)
+					}
+				}
+			}
+			continue
+		}
+
+		K8sResources = append(K8sResources, manifest)
+	}
+
+	return K8sResources, nil
+}
+
+// NewPatternFileFromK8sManifest converts Kubernetes manifests into a Meshery design pattern file.
+// It extracts all resources, processes each resource to generate component
+// definitions, and combines them into a design pattern. If `ignoreErrors` is true, it skips errors
+// during the creation of individual component definitions.
+//
+// Note: If modified, make sure this function always returns a meshkit error
+func NewPatternFileFromK8sManifest(data string, fileName string, ignoreErrors bool, reg *registry.RegistryManager) (pattern.PatternFile, error) {
+	if fileName == "" {
+		fileName = "Autogenerated"
+	}
+
+	pattern := pattern.PatternFile{
+		SchemaVersion: v1beta1.DesignSchemaVersion,
+		Name:          fileName,
+		Components:    []*component.ComponentDefinition{},
+	}
+
+	allK8sResources, err := getAllK8sResources(data, ignoreErrors)
+	if err != nil {
+		return pattern, err
+	}
+
+	if len(allK8sResources) == 0 {
+		return pattern, ErrParseK8sManifest(fmt.Errorf("Kubernetes manifest is empty"))
+	}
+
+	// Process each resource
+	for _, manifest := range allK8sResources {
+
```

Comment:
i think we could process list kinds here too , to prevent too whole iterations of the manifest file .

Reply:
not that big of a deal ,if separating it helps with code readability and separation of logic .